packages/hierarchy-builder/src/hierarchy-builder/internal/operators/grouping/PropertiesGrouping.ts

```diff
@@ -0,0 +1,332 @@
+}
+
+/** @internal */
+export type PreviousPropertiesGroupingInfo = Array<{ fullClassName: string; propertyGroup: Omit<PropertyGroup, "propertyValue"> }>;
+
+/** @internal */
+export async function createPropertyGroups(
+  metadata: IMetadataProvider,
+  nodes: ProcessedInstanceHierarchyNode[],
+  propertyInfo: PropertyGroupInfo,
+): Promise<GroupingHandlerResult> {
+  const groupings: PropertyGroupingInformation = { ungrouped: [], grouped: new Map() };
+  for (const node of nodes) {
+    const byProperties = node.processingParams?.grouping?.byProperties;
+    if (!byProperties) {
+      groupings.ungrouped.push(node);
+      continue;
+    }
+    if (!(await shouldCreatePropertyGroup(metadata, propertyInfo, byProperties, node.key.instanceKeys[0].className))) {
+      groupings.ungrouped.push(node);
+      continue;
+    }
+    const currentProperty = byProperties.propertyGroups[propertyInfo.previousPropertiesGroupingInfo.length];
+    const propertyToGroup = {
+      propertyName: currentProperty.propertyName,
+      fullClassName: byProperties.fullClassName,
+    };
+
+    if (currentProperty.ranges) {
+      if (typeof currentProperty.propertyValue !== "number") {
+        addGroupingToMap(
+          groupings.grouped,
+          `${currentProperty.propertyName}:Other`,
+          {
+            label: "Other",
+            ...propertyToGroup,
+          },
+          node,
+        );
+        continue;
+      }
+      const propValue = currentProperty.propertyValue;
+      const matchingRange = currentProperty.ranges.find((range) => propValue >= range.fromValue && propValue <= range.toValue);
+
+      if (matchingRange) {
+        const rangeLabel = matchingRange.rangeLabel ?? `${matchingRange.fromValue} - ${matchingRange.toValue}`;
+        addGroupingToMap(
+          groupings.grouped,
+          `${currentProperty.propertyName}:${matchingRange.fromValue} - ${matchingRange.toValue}(${rangeLabel})`,
+          {
+            label: rangeLabel,
+            ...propertyToGroup,
+            extraValues: {
+              fromValue: matchingRange.fromValue,
+              toValue: matchingRange.toValue,
+            },
+          },
+          node,
+        );
+        continue;
+      }
+    }
+
+    const propertyClass = await getClass(metadata, byProperties.fullClassName);
+    const property = await propertyClass.getProperty(currentProperty.propertyName);
+    if (!property?.isPrimitive()) {
+      addGroupingToMap(
+        groupings.grouped,
+        `${currentProperty.propertyName}:Other`,
+        {
+          label: "Other",
+          ...propertyToGroup,
+        },
+        node,
+      );
+      continue;
+    }
+    const part = {
+      type: property.primitiveType,
+      extendedType: property.extendedTypeName,
+      koqName: (await property.kindOfQuantity)?.fullName,
+      value: currentProperty.propertyValue,
+    } as TypedPrimitiveValue;
+    const valueFormatter = createDefaultValueFormatter();
```

Comment:
should use the formatter from `HierarchyProvider`

Reply:
Changed